ramips: add support for Netgear R6260 & R6850

achterin opened the pull request:
ramips: add support for Netgear R6260 and R6850

As Netgear uses the same image for R6260, R6350 & R6850
we can merge device tree files and generate separate
images for each device.

ramips: add factory images for Netgear R6260, R6350 and R6850

This adds factory image generation for all three
devices. These images can be flashed via WebUI
for easy installation.
    
Tested on Netgear R6260 using V1.1.0.52
Thanks to David Bauer for the inspiration

achterin commented:
I've updated my PR because of a discovery made by @nicefile: Netgear is using the same image for R6260, R6350 & R6850. Therefore I've merged the dts files and added separate image generation for all three devices.
I've also added factory image generation based on the work of @blocktrron. These images can be flashed via vendor UI.

blocktrron commented on target/linux/ramips/base-files/etc/board.d/01_leds:
```diff
@@ -293,7 +293,9 @@ netgear,r6120)
+netgear,r6850|\
 netgear,wndr3700-v5)
 	ucidef_set_led_switch "wan" "wan" "$boardname:green:wan" "switch0" "0x10"
 	set_wifi_led "$boardname:green:wifi"
```
New devices should have it's wifi led trigger defined in the DTS (given the driver supports it). This should be the case for mt76 supported chips.

achterin replied:
changed that part. is it okay now?

blocktrron replied:
See my other comment.

blocktrron commented on target/linux/ramips/dts/mt7621_netgear_r6260_r6350_r6850.dtsi:
```diff
@@ -0,0 +1,158 @@
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
```
MediaTek MT7621 should support interrupt-driven gpio-button support.  Can you verify this works on your board?

achterin replied:
works, changed.

blocktrron commented on target/linux/ramips/dts/mt7621_netgear_r6260_r6350_r6850.dtsi:
```diff
@@ -0,0 +1,158 @@
+		};
+
+		led_wifi: wifi {
+			gpios = <&gpio0 16 GPIO_ACTIVE_LOW>;
```
Add `linux,default-trigger = "phy0tpt"` here.

achterin replied:
added

blocktrron replied:
Okay, this is completely my bad - it is not `phy0tpt`, but `phy1tpt` for the 2.4 GHz radio. We can't trigger the LED using both radios, so the 2.4GHz one should be the default.

The 2.4 GHz radio was also triggered before via userspace, as only the MT7612 driver was present.

achterin replied:
no worries, phy0 is the 2.4GHz radio: https://pastebin.com/vfNFaGgT

blocktrron commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -409,24 +409,45 @@ define Device/netgear_r6220
-  DEVICE_MODEL := R6350
   DEVICE_PACKAGES := \
-	kmod-mt7603 kmod-usb3 kmod-usb-ledtrig-usbport wpad-basic
+	kmod-mt7603 kmod-mt7615e kmod-usb3 kmod-usb-ledtrig-usbport wpad-basic
```
The addition of the mt7615 driver for the R6350 is not mentioned in your commits. Maybe you can add this as the first commit of your patchset?

achterin replied:
fixed

adrianschmutzler commented:
Based on your current MAC assignment, it seems very likely that you will have additional MAC addresses in the factory partition (to be used for wan/lan). Easiest way to check on that would be to do a hexdump of the entire factory partition and then search this output for occurrences of a Mac hexpair not supposed to change.

achterin commented:
@adrianschmutzler Current MAC assignment matches OEM, so i dont know if your assumption is true but i can check later. But if there are any, to which interface should these get assigned to?

achterin committed:
ramips: add mt7615e support to Netgear R6350

Signed-off-by: Christoph Krapp <achterin@googlemail.com>

achterin committed:
ramips: add support for Netgear R6260 and R6850

As Netgear uses the same image for R6260, R6350 & R6850
we can merge device tree files and generate separate
images for each device.

Signed-off-by: Christoph Krapp <achterin@googlemail.com>

adrianschmutzler commented:
I'm on travel, will give a more detailed explanation at the weekend...

adrianschmutzler commented:
(I suspect that you will find two additional addresses, one for wan and one for lan, of which one is the same as 0x4)

achterin commented:
@adrianschmutzler negative. there are only 2 MACs in the factory partition, one at 0x4 and the other one at 0x8004.

achterin committed:
ramips: add factory images for Netgear R6260, R6350 and R6850

This adds factory image generation for all three
devices. These images can be flashed via WebUI
for easy installation.

Tested on Netgear R6260 using V1.1.0.52
Thanks to David Bauer for the inspiration

Signed-off-by: Christoph Krapp <achterin@googlemail.com>

blocktrron commented on target/linux/ramips/base-files/etc/board.d/01_leds:
```diff
@@ -293,11 +293,16 @@ netgear,r6120)
+netgear,r6350|\
+netgear,r6850)
+	ucidef_set_led_switch "wan" "wan" "$boardname:green:wan" "switch0" "0x10"
+	ucidef_set_led_wlan "wifi" "wifi" "$boardname:green:wifi" "phy0tpt"
```
This userspace WiFi-LED definition can be dropped altogether, as the trigger is defined in the DTSI by default. The user can still change the LEDs behavior from userspace, however this redundancy can be dropped.

achterin replied:
yeah, but it will be missing in /etc/config/system and thus luci by default. i dont think thats very userfriendly. if you insist on dropping that line, i'll do so.

blocktrron replied:
This was topic in a ML thread some time ago as i've noted the same concern back then. As the board has the WiFi LED explicitl set, it's a property of the device (the user can change later by creating his/her own copnfiguration.

So yes, please drop it.

achterin replied:
Allright, didnt know that. Thanks for the explanation. Will keep that in mind for the next device :)

blocktrron commented:
@achterin As I'm not sure if GitHub sends out notifications for this just a heads up: https://github.com/openwrt/openwrt/pull/2358#discussion_r319265726

Excuse my derpiness :panda_face:

achterin commented:
@blocktrron all good, phy0 is the 2.4GHz radio.

blocktrron commented:
Thanks, merged into my staging tree: https://git.openwrt.org/?p=openwrt/staging/blocktrron.git;a=summary

blocktrron commented:
I'm reopening this. I've tested the changes in my staging tree with my R6220 and discovered that the switchport numbering in LuCi is wrong.

This is not an issue with your PR per se, but can you confirm the switchport numbering is wrong on your device too? Photos of the R6350 casing suggest it has the same arrangement / order as the R6220, thats why I'm asking.

achterin commented:
you are right, i can confirm that the switchport numbering is reversed. sorry about that, i didnt check as i thought the r6350 commit took already care of that. stupid me. thanks for checking!

blocktrron commented:
Thanks for confirming this.

Looking at the MAC assignment which differs from the R6120: Can you also confirm, that the MAC assignment for the wired interfaces (LAN/WAN) is inverted compared to the Netgear firmware? It is the case for my R6220 :disappointed:

achterin commented:
You are partially correct this time. This is the assignment using OEM:
LAN ends on :07 (as noted on the case)
WAN ends on :09
So yes, they are reversed but also +2 instead of +1 apart.
I should really check everything again, even if i am only adding support for a clone of an already supported board. Sorry.

blocktrron commented:
Okay, i will incorporate this in my fix commit (as i have to fix it for the R6220 anyway).

Can you build from my staging tree again? This should hopefully address (:smile:) all issues:

https://git.openwrt.org/?p=openwrt/staging/blocktrron.git;a=commit;h=9a4fd256a72256c568b08666948cf57319a61a2b

achterin commented:
I can confirm: all issues are fixed. Thanks for your work!

blocktrron commented:
Thanks, finally merged :+1:

adrianschmutzler commented:
> LAN ends on :07 (as noted on the case)

So, for which of the mentioned devices is LAN MAC (0x4) = case MAC?

achterin commented:
LAN & WLAN 2.4 GHz = Case MAC

adrianschmutzler commented:
@achterin For R6260, R6850 and 6350? (as they are similar anyway?)

achterin commented:
@adrianschmutzler yes, all three have the same mac setup
